File: src/test/java/net/pterodactylus/sone/web/CreateSonePageTest.java
package net.pterodactylus.sone.web;

import static net.pterodactylus.sone.web.WebTestUtils.redirectsTo;
import static net.pterodactylus.util.web.Method.POST;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.contains;
import static org.hamcrest.Matchers.is;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import java.util.Arrays;
import java.util.Collection;
import java.util.HashSet;

import net.pterodactylus.sone.data.Profile;
import net.pterodactylus.sone.data.Sone;
import net.pterodactylus.sone.freenet.wot.OwnIdentity;

import org.junit.Test;
import org.mockito.invocation.InvocationOnMock;
import org.mockito.stubbing.Answer;

/**
 * Unit test for {@link CreateSonePage}.
 *
 * @author <a href="mailto:bombe@pterodactylus.net">David ‘Bombe’ Roden</a>
 */
public class CreateSonePageTest extends WebPageTest {

	private final CreateSonePage page = new CreateSonePage(template, webInterface);
	private final Sone[] localSones = { createSone("local-sone1"), createSone("local-sone2"), createSone("local-sone3") };
	private final OwnIdentity[] ownIdentities = {
			createOwnIdentity("own-id-1", "Sone"),
			createOwnIdentity("own-id-2", "Test", "Foo"),
			createOwnIdentity("own-id-3"),
			createOwnIdentity("own-id-4", "Sone")
	};

	@Test
	public void pageReturnsCorrectPath() {
		assertThat(page.getPath(), is("createSone.html"));
	}

	@Test
	@SuppressWarnings("unchecked")
	public void getRequestStoresListOfIdentitiesInTemplateContext() throws Exception {
		addDefaultLocalSones();
		addDefaultOwnIdentities();
		page.processTemplate(freenetRequest, templateContext);
		assertThat((Collection<Sone>) templateContext.get("sones"), contains(localSones[0], localSones[1], localSones[2]));
		assertThat((Collection<OwnIdentity>) templateContext.get("identitiesWithoutSone"), contains(ownIdentities[1], ownIdentities[2]));
	}

	private void addDefaultLocalSones() {
		addLocalSone("local-sone3", localSones[2]);
		addLocalSone("local-sone1", localSones[0]);
		addLocalSone("local-sone2", localSones[1]);
	}

	private void addDefaultOwnIdentities() {
		addOwnIdentity(ownIdentities[2]);
		addOwnIdentity(ownIdentities[0]);
		addOwnIdentity(ownIdentities[3]);
		addOwnIdentity(ownIdentities[1]);
	}

	private Sone createSone(String id) {
		Sone sone = mock(Sone.class);
		when(sone.getId()).thenReturn(id);
		when(sone.getProfile()).thenReturn(new Profile(sone));
		return sone;
	}

	private OwnIdentity createOwnIdentity(String id, final String... contexts) {
		OwnIdentity ownIdentity = mock(OwnIdentity.class);
		when(ownIdentity.getId()).thenReturn(id);
		when(ownIdentity.getNickname()).thenReturn(id);
		when(ownIdentity.getContexts()).thenReturn(new HashSet<>(Arrays.asList(contexts)));
		when(ownIdentity.hasContext(anyString())).thenAnswer(new Answer<Boolean>() {
			@Override
			public Boolean answer(InvocationOnMock invocation) throws Throwable {
				return Arrays.asList(contexts).contains(invocation.<String>getArgument(0));
			}
		});
		return ownIdentity;
	}

	@Test
	public void soneIsCreatedAndLoggedIn() throws Exception {
		addDefaultLocalSones();
		addDefaultOwnIdentities();
		addHttpRequestParameter("identity", "own-id-3");
		request("", POST);
		Sone newSone = mock(Sone.class);
		when(core.createSone(ownIdentities[2])).thenReturn(newSone);
		expectedException.expect(redirectsTo("index.html"));
		try {
			page.processTemplate(freenetRequest, templateContext);
		} finally {
			verify(core).createSone(ownIdentities[2]);
			verify(webInterface).setCurrentSone(toadletContext, newSone);
		}
	}

	@Test
	public void onInvalidIdentityIdFlagIsStoredInTemplateContext() throws Exception {
		addDefaultLocalSones();
		addDefaultOwnIdentities();
		addHttpRequestParameter("identity", "own-id-invalid");
		request("", POST);
		page.processTemplate(freenetRequest, templateContext);
		assertThat(((Boolean) templateContext.get("errorNoIdentity")), is(true));
	}

	@Test
	public void ifSoneIsNotCreatedUserIsStillRedirectedToIndex() throws Exception {
		addDefaultLocalSones();
		addDefaultOwnIdentities();
		addHttpRequestParameter("identity", "own-id-3");
		request("", POST);
		when(core.createSone(ownIdentities[2])).thenReturn(null);
		expectedException.expect(redirectsTo("index.html"));
		try {
			page.processTemplate(freenetRequest, templateContext);
		} finally {
			verify(core).createSone(ownIdentities[2]);
			verify(webInterface).setCurrentSone(toadletContext, null);
		}
	}

	@Test
	public void doNotShowCreateSoneInMenuIfFullAccessRequiredButClientHasNoFullAccess() {
		when(core.getPreferences().isRequireFullAccess()).thenReturn(true);
		when(toadletContext.isAllowedFullAccess()).thenReturn(false);
		assertThat(page.isEnabled(toadletContext), is(false));
	}

	@Test
	public void showCreateSoneInMenuIfNotLoggedInAndClientHasFullAccess() {
		when(core.getPreferences().isRequireFullAccess()).thenReturn(true);
		when(toadletContext.isAllowedFullAccess()).thenReturn(true);
		unsetCurrentSone();
		assertThat(page.isEnabled(toadletContext), is(true));
	}

	@Test
	public void showCreateSoneInMenuIfNotLoggedIn() {
		unsetCurrentSone();
		assertThat(page.isEnabled(toadletContext), is(true));
	}

	@Test
	public void showCreateSoneInMenuIfLoggedInAndASingleSoneExists() {
		addLocalSone("local-sone", mock(Sone.class));
		assertThat(page.isEnabled(toadletContext), is(true));
	}

	@Test
	public void doNotShowCreateSoneInMenuIfLoggedInAndMoreLocalSonesExists() {
		addLocalSone("local-sone1", mock(Sone.class));
		addLocalSone("local-sone2", mock(Sone.class));
		assertThat(page.isEnabled(toadletContext), is(false));
	}

}

File: src/test/kotlin/net/pterodactylus/sone/web/CreateSonePageTest.kt
package net.pterodactylus.sone.web

import net.pterodactylus.sone.data.Profile
import net.pterodactylus.sone.data.Sone
import net.pterodactylus.sone.freenet.wot.OwnIdentity
import net.pterodactylus.sone.test.mock
import net.pterodactylus.sone.test.whenever
import net.pterodactylus.util.web.Method.POST
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.contains
import org.hamcrest.Matchers.equalTo
import org.junit.Test
import org.mockito.ArgumentMatchers.anyString
import org.mockito.Mockito.verify

/**
 * Unit test for [CreateSonePage].
 */
class CreateSonePageTest: WebPageTest() {

	private val page = CreateSonePage(template, webInterface)
	override fun getPage() = page

	private val localSones_ = listOf(
			createSone("local-sone1"),
			createSone("local-sone2"),
			createSone("local-sone3")
	)

	private fun createSone(id: String) = mock<Sone>().apply {
		whenever(this.id).thenReturn(id)
		whenever(profile).thenReturn(Profile(this))
	}

	private val ownIdentities_ = listOf(
			createOwnIdentity("own-id-1", "Sone"),
			createOwnIdentity("own-id-2", "Test", "Foo"),
			createOwnIdentity("own-id-3"),
			createOwnIdentity("own-id-4", "Sone")
	)

	private fun createOwnIdentity(id: String, vararg contexts: String) = mock<OwnIdentity>().apply {
		whenever(this.id).thenReturn(id)
		whenever(this.nickname).thenReturn(id)
		whenever(this.contexts).thenReturn(contexts.toSet())
		whenever(this.hasContext(anyString())).thenAnswer { invocation -> invocation.getArgument<String>(0) in contexts }
	}

	@Test
	fun `page returns correct path`() {
		assertThat(page.path, equalTo("createSone.html"))
	}

	@Test
	fun `page does not require login`() {
		assertThat(page.requiresLogin(), equalTo(false))
	}

	private fun addExistingSones() {
		listOf(2, 0, 1).map { localSones_[it] }.forEach { addLocalSone(it.id, it) }
	}

	@Test
	@Suppress("UNCHECKED_CAST")
	fun `get request stores sorted list of local sones in template context`() {
		addExistingSones()
		page.processTemplate(freenetRequest, templateContext)
		assertThat(templateContext["sones"] as Collection<Sone>, contains(localSones_[0], localSones_[1], localSones_[2]))
	}

	private fun addExistingOwnIdentities() {
		listOf(2, 0, 3, 1).map { ownIdentities_[it] }.forEach { addOwnIdentity(it) }
	}

	@Test
	@Suppress("UNCHECKED_CAST")
	fun `get request stores sorted sones without sone context in the template context`() {
		addExistingOwnIdentities()
		page.processTemplate(freenetRequest, templateContext)
		assertThat(templateContext["identitiesWithoutSone"] as Collection<OwnIdentity>, contains(ownIdentities_[1], ownIdentities_[2]))
	}

	@Test
	fun `sone is created and logged in`() {
		addExistingOwnIdentities()
		request("", POST)
		addHttpRequestParameter("identity", "own-id-3")
		val newSone = mock<Sone>()
		whenever(core.createSone(ownIdentities_[2])).thenReturn(newSone)
		verifyRedirect("index.html") {
			verify(webInterface).setCurrentSone(toadletContext, newSone)
		}
	}

	@Test
	fun `on invalid identity id a flag is set in the template context`() {
		request("", POST)
		addHttpRequestParameter("identity", "own-id-3")
		page.processTemplate(freenetRequest, templateContext)
		assertThat(templateContext["errorNoIdentity"], equalTo<Any>(true))
	}

	@Test
	fun `if sone is not created user is still redirected to index`() {
		addExistingOwnIdentities()
		request("", POST)
		addHttpRequestParameter("identity", "own-id-3")
		whenever(core.createSone(ownIdentities_[2])).thenReturn(null)
		verifyRedirect("index.html") {
			verify(core).createSone(ownIdentities_[2])
			verify(webInterface).setCurrentSone(toadletContext, null)
		}
	}

	@Test
	fun `create sone is not shown in menu if full access is required but client doesn’t have full access`() {
		core.preferences.isRequireFullAccess = true
		assertThat(page.isEnabled(toadletContext), equalTo(false))
	}

	@Test
	fun `create sone is shown in menu if no sone is logged in`() {
		unsetCurrentSone()
		assertThat(page.isEnabled(toadletContext), equalTo(true))
	}

	@Test
	fun `create sone is shown in menu if a single sone exists`() {
		addLocalSone("local-sone", localSones_[0])
		assertThat(page.isEnabled(toadletContext), equalTo(true))
	}

	@Test
	fun `create sone is not shown in menu if more than one sone exists`() {
		addLocalSone("local-sone1", localSones_[0])
		addLocalSone("local-sone2", localSones_[1])
		assertThat(page.isEnabled(toadletContext), equalTo(false))
	}

	@Test
	fun `create sone is shown in menu if no sone is logged in and client has full access`() {
		core.preferences.isRequireFullAccess = true
		whenever(toadletContext.isAllowedFullAccess).thenReturn(true)
		unsetCurrentSone()
		assertThat(page.isEnabled(toadletContext), equalTo(true))
	}

}
